### ISSUE TITLE
Elasticsearch plugins can no longer define sibiling pipeline aggregator

### ISSUE THREAD

**Issue by tools4origins** [elastic/elasticsearch#42493]
**Elasticsearch version** : v7.0.1, v7.1.0
**Plugins installed**: []
**JVM version** (`java -version`): Any
**OS version** (`uname -a` if on a Unix-like system): Any

**Description of the problem including expected versus actual behavior**:
Since the PR https://github.com/elastic/elasticsearch/pull/34658 it no longer is possible to define a custom sibling pipeline aggregation as all SiblingPipelineAggregator constructors are package-private.

A plugin we developed add such metrics in ES 5 and ES 6, unfortunately because of this change we cannot adapt the code for ES 7.

My guess is that the constructor accessibility change was done without having plugin development in mind (tagging you @polyfractal as, as the author of the PR, you may be able to confirm that?).

Could it be set back to public?

Similar limitations are caused by package-private constructors and final methods in the org.elasticsearch.search.aggregations module. Those one can be bypassed by duplicating code instead of inheriting.

The issue with SiblingPipelineAggregator is that there is [a check in aggregation phase](https://github.com/elastic/elasticsearch/blob/c1de8c29db228bb633f6766684874ed3321d7200/server/src/main/java/org/elasticsearch/search/aggregations/AggregationPhase.java#L138
) that specifically check `pipelineAggregator instanceof SiblingPipelineAggregator` so we must inherit SiblingPipelineAggregator to define a sibling pipeline aggregator. But we no longer can :(.

We would highly appreciate a accessibility change for SiblingPipelineAggregator constructors, I assume for other constructors and final method the design has more reasons to exist and it doesn't restrict us even though it make our code a bit uglier.

**Steps to reproduce**:

With this class defined:
```
package com.organization.customplugin.elasticsearch.search.aggregations.pipeline;

import org.elasticsearch.search.aggregations.pipeline.SiblingPipelineAggregator;

import java.util.Map;


public abstract class PublicSiblingPipelineAggregator extends SiblingPipelineAggregator {
    PublicSiblingPipelineAggregator(String name, String[] bucketsPaths, Map<String, Object> metaData) {
        super(name, bucketsPaths, metaData);
    }
}
```

This compilation error is raised:
`error: PublicSiblingPipelineAggregator(String,String[],Map<String,Object>) is not public in PublicSiblingPipelineAggregator; cannot be accessed from outside package
`

Dependencies:

```
    <dependencies>
        <dependency>
            <groupId>org.elasticsearch</groupId>
            <artifactId>elasticsearch</artifactId>
            <version>7.1.0</version>
            <scope>compile</scope>
        </dependency>
    </dependencies>
```

**Comment by polyfractal** [comment]
Hey @tools4origins, sorry about this!  Wasn't intended to break plugins, just an oversight.  We were collapsing the agg packages so that various methods (getters on the `Internal*` classes, etc) that should be private could actually become package-private, and just to reduce the package complexity.  Making something like `SiblingPipelineAggregator` inaccessible to plugin authors was not intended.

I'll take a closer look this week, but I don't think it will be a problem to promote `SiblingPipelineAggregator` back to public.

**Comment by polyfractal** [comment]
@tools4origins I just merged a PR to make the ctor protected, so I think this should be accessible for plugin authors now.

**Comment by Pnchbllz** [comment]
Hello @polyfractal, 

The change seems to have been merged in master but was not released yet:
master: https://github.com/elastic/elasticsearch/blame/master/server/src/main/java/org/elasticsearch/search/aggregations/pipeline/SiblingPipelineAggregator.java
v7.4.0: https://github.com/elastic/elasticsearch/blame/v7.4.0/server/src/main/java/org/elasticsearch/search/aggregations/pipeline/SiblingPipelineAggregator.java

We are also blocked in our migration by this regression, do you know where it stands in the roadmap for release?

**Comment by polyfractal** [comment]
Oof, I'm not sure what happened when I was merging but it looks like it just didn't get backported.  Let me ping a few folks and see if we can backport this to a bugfix release so that it is available sooner than the next minor.  

Sorry for the delay!  Entirely my fault :(

**Comment by Pnchbllz** [comment]
Hello,
Thank you very much for you answers and actions.
Do you have an idea of when it will be released  ?

**Comment by polyfractal** [comment]
Hi @Pnchbllz, I just merged this into `7.4` (https://github.com/elastic/elasticsearch/commit/d0fe074cf34a411dd7abcaab43433fe7f3937edc) and `7.x` (https://github.com/elastic/elasticsearch/commit/2de3411c9cec763bf8c6b3f1f030cb60258c714f)

I don't have a date for when it will release, but the next bugfix for the 7.4 branch (`7.4.1`) will contain this change.  Apologies again for the oversight when it first merged!

**Comment by tools4origins** [comment]
Thanks a lot! We're looking forward for the next release then
